### PR TITLE
Enable streaming / CDF streaming reads on column mapping enabled tables with fewer limitations

### PR DESCRIPTION
**Pull request by jackierwzhang**
## Description
Resolves https://github.com/delta-io/delta/issues/1357

As streaming uses the latest schema to read historical data batches and column mapping schema changes (e.g. rename/drop column) can cause latest schema to diverge, we decided to temporarily completely block streaming read on column mapping tables before.

As a close follow up in this PR, we think it is at least possible to enable the following use cases:

Read from a column mapping table without rename or drop column operations.
Upgrade to column mapping tables.
Existing compatible schema change operations such as ADD COLUMN.

## How was this patch tested?
New unit tests.

## Does this PR introduce _any_ user-facing changes?
No

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala [hunk 1]
```diff
@@ -522,15 +522,17 @@ trait DeltaColumnMappingBase extends DeltaLogging {
   def isColumnMappingReadCompatible(newMetadata: Metadata, oldMetadata: Metadata): Boolean = {
     val (oldMode, newMode) = (oldMetadata.columnMappingMode, newMetadata.columnMappingMode)
     if (oldMode != NoMapping && newMode != NoMapping) {
+      require(oldMode == newMode, "changing mode is not supported")
```
Comment (scottsand-db) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala:525]: do you have a test for this?

File: core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala [hunk 2]
```diff
@@ -438,6 +541,9 @@ case class DeltaSource(
     initialState.iterator()
   }
 
+  /**
+   * Narrow-waist for generating snapshot from Delta Log within Delta Source
```
Comment (scottsand-db) [inline core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:545]: what does narrow-waist mean in this context?

File: core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSourceCDCSupport.scala [hunk 3]
```diff
@@ -262,17 +270,18 @@ trait DeltaSourceCDCSupport { self: DeltaSource =>
...
+            val cdcSchema = CDCReader.cdcReadSchema(m.schema)
+            if (!SchemaUtils.isReadCompatible(cdcSchema, schema)) {
+              throw DeltaErrors.schemaChangedException(schema, cdcSchema, false)
+            }
           }
           false
         case protocol: Protocol =>
           deltaLog.protocolRead(protocol)
           false
-        case _: SetTransaction | _: CommitInfo =>
```
Comment (scottsand-db) [inline core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSourceCDCSupport.scala:273]: why this change?

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala (hunk 1 above)
Comment (jackierwzhang) [reply core/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala:525]: This is just defense in depth (or me being paranoid :)), we should've already blocked column mapping changing modes, e.g. [here](https://github.com/delta-io/delta/blob/master/core/src/main/scala/org/apache/spark/sql/delta/DeltaColumnMapping.scala#L119)

File: core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala (hunk 2 above)
Comment (jackierwzhang) [reply core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala:545]: This is just to note, because I believe this function is the ultimate function for generating snapshots across all places of DeltaSource, so any new usage should adopt this util.

File: core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSourceCDCSupport.scala (hunk 3 above)
Comment (jackierwzhang) [reply core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSourceCDCSupport.scala]: Sorry this was unnecessary, reverted.

**Comment by tnk-dev** [comment]
Hey Delta folks,

since the only way to allow **renaming** and **dropping** columns in delta **without rewriting everything** is to enable `columnMapping`, which **collides with Change Data Feed** resulting in 
`com.databricks.sql.transaction.tahoe.DeltaColumnMappingUnsupportedSchemaIncompatibleException: Change Data Feed (CDF) read is not supported on tables with column mapping schema changes (e.g. rename or drop)...`
and can only be solved with `spark.databricks.delta.changeDataFeed.unsafeBatchReadOnIncompatibleSchemaChanges.enabled`, we as a company are kinda **scared** of enabling this option and are not sure at all about the **ramifications** of enabling `unsafeBatchReadOnIncompatibleSchemaChanges`. Plus we don't really like the 2 character partitioning convention in our Datalake (S3), that is replaced by the `columnMapping` option.

So my question is:
**What is best practice here according to Delta regarding above mentioned pain points?**

@dennyglee
